### ISSUE TITLE
testapp.c: undefined reference to 'cache_error'

### ISSUE THREAD

**Issue by fscm** [memcached/memcached#830]
Since `cache_error` variable is only declared on the `cache.h` file, shouldn't the same variable be defined somewhere in the `test app.c` file (before the `cache_redzone_test` function or in the beginning of this function)?

I am getting the error "testapp.c: undefined reference to 'cache_error'" when running "make" on CentOS. Defining the variable (```int cache_error;```) in the beginning of the `cache_redzone_test` function (after the `#ifndef HAVE_UMEM_H` line) solved this error.

I am using the source package version 1.6.12 downloaded from the 'memcached.org' page.

**Comment by dormando** [comment]
I'm not sure how this error is happening. `cache_error` is defined in the .c file for the same circumstances it gets defined in `cache.h`, ifndef NDEBUG (only for the debug mode builds). Is your build system building testapp with NDEBUG somehow?

Sorry super late reply. :/

**Comment by fscm** [comment]
Yes, you are right. The 'NDEBUG' flag was being set.

That means that this error only happens when that flag is set. For this not to happen, the `test_app.c` file should maybe include the following code (like the `cache.c` file):

```
#ifndef NDEBUG
int cache_error = 0;
#endif
```

Is this still a bug or intended purpose?

**Comment by dormando** [comment]
Can you explain why you're intentionally passing the NDEBUG flag while building testapp's dependencies? If I'm understanding correctly and that's what's happening here.

**Comment by fscm** [comment]
The "NDEBUG" flag was used as part of a process to create the smallest possible static version of the `memcached` app.
The `testapp` app is just part of the default Makefile action... I did not _ intentionally _ intended for the `testapp` to be built.

**Comment by dormando** [comment]
The main binary should already be compiled with NDEBUG; that flag is only omitted for the `memcached-debug` and testapp binaries. Or did I break that somehow?

**Comment by fscm** [comment]
No, the NDEBUG flag is indeed being used in the compilation of the main binary.

I did not notices the usage of the NDEBUG flag on the default build process. I used it as part of a set of flags for the entire process (the creation of a static version of the memcached app) that is why I have asked previously if this was still a bug or an intended behaviour.

It is valid for me to not consider this a bug if the break of the building process when using that flag is intended. Likewise, it is also valid for me to consider this a bug in the way that the usage of a flag, that is already being used by default, should not break the building process. I will leave it up to you (your app you call :)

...And thank you for the help on better understanding this behaviour.

**Comment by dormando** [comment]
Sorry I keep having to come back to this at a low priority. The very first line of testapp.c undefine's NDEBUG, but the problem is in NDEBUG creeping into its dependency objects. I'm not sure I have the bandwidth to make a decision either way right now, so I'll leave this issue open and come back to it the next time I look at the build code. If you want to build and run the tests I'd say throwing the NDEBUG flag around is going to absolutely break the test suite unless we explicitly defend for it. At best it'd make the tests pointless since a lot of the verification code would be elided.

So short term; don't override NDEBUG. In the future I may decide to defend against having it supplied for the test builds since those don't work with it.

**Comment by dormando** [comment]
I'm going to close this out... testapp is slated for eventual removal and the case here seems uncommon.
